usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -489,6 +491,394 @@ static CK_RV cleanse_attribute(TEMPLATE *template,
+                                       OBJECT *obj,
+                                       CK_KEY_TYPE keytype)
+{
+    ep11_private_data_t *ep11_data = tokdata->private_data;
+    CK_RV rc;
+    CK_BBOOL trusted, encrypt, decrypt, wrap, unwrap, sign, sign_recover,
+             verify, verify_recover, derive, extractable, local,
+             never_extractable, modifiable, wrap_with_trusted;
+    CK_ULONG valuelen;
+    CK_KEY_TYPE template_keytype;
+    CK_ATTRIBUTE attrs[] = {
+        {CKA_TRUSTED,           &trusted,           sizeof(trusted)},
+        {CKA_ENCRYPT,           &encrypt,           sizeof(encrypt)},
+        {CKA_DECRYPT,           &decrypt,           sizeof(decrypt)},
+        {CKA_WRAP,              &wrap,              sizeof(wrap)},
+        {CKA_UNWRAP,            &unwrap,            sizeof(unwrap)},
+        {CKA_SIGN,              &sign,              sizeof(sign)},
+        {CKA_SIGN_RECOVER,      &sign_recover,      sizeof(sign_recover)},
+        {CKA_VERIFY,            &verify,            sizeof(verify)},
+        {CKA_VERIFY_RECOVER,    &verify_recover,    sizeof(verify_recover)},
+        {CKA_DERIVE,            &derive,            sizeof(derive)},
+        {CKA_EXTRACTABLE,       &extractable,       sizeof(extractable)},
+        {CKA_LOCAL,             &local,             sizeof(local)},
+        {CKA_NEVER_EXTRACTABLE, &never_extractable, sizeof(extractable)},
+        {CKA_MODIFIABLE,        &modifiable,        sizeof(modifiable)},
+        {CKA_WRAP_WITH_TRUSTED, &wrap_with_trusted, sizeof(wrap_with_trusted)},
+        {CKA_KEY_TYPE,          &template_keytype,  sizeof(template_keytype)},
+        {CKA_VALUE_LEN,         &valuelen,          sizeof(valuelen)},
+    };
+    CK_ULONG i;
+    CK_ATTRIBUTE *attr;
+    CK_BBOOL cktrue = TRUE;
+
+    rc = dll_m_GetAttributeValue(blob, blob_len, attrs,
+                                 sizeof(attrs) / sizeof(CK_ATTRIBUTE),
+                                 ep11_data->target);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Retrieving attributes from AB unwrapped key failed, rc=0x%lx\n",
+                    rc);
+        return rc;
+    }
+    if (template_keytype != keytype) {
+        TRACE_ERROR("Template specifies different key type than the AB key (0x%08lx vs 0x%08lx)\n",
+                    template_keytype, keytype);
+        return CKR_TEMPLATE_INCONSISTENT;
+    }
+    for (i = 0; i < sizeof(attrs) / sizeof(CK_ATTRIBUTE); ++i) {
+        rc = build_attribute(attrs[i].type, attrs[i].pValue,
+                             attrs[i].ulValueLen, &attr);
+        if (rc != CKR_OK) {
+            TRACE_ERROR("failed to build attribute (rc=0x%lx)\n", rc);
+            return rc;
+        }
+        rc = template_update_attribute(obj->template, attr);
+        if (rc != CKR_OK) {
+            TRACE_ERROR("%s template_update_attribute failed with rc=0x%lx\n",
+                        __func__, rc);
+            return rc;
```

Comment:
If this has failed, then you should free attr. build_attribute allocates the attr, and template_update_attribute adds it to the list. But if template_update_attribute() fails, then attr has not been added, and thus must be freed.

Reply:
Good find.  Fixed.